rcl/src/rcl/wait.c

```diff
@@ -541,54 +541,123 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
-      // TODO(sloretz) fix spurious wake-ups on ROS_TIME timers with ROS_TIME enabled
-      int64_t timer_timeout = INT64_MAX;
-      rcl_ret_t ret = rcl_timer_get_time_until_next_call(wait_set->timers[i], &timer_timeout);
+
+      if (clock->type == RCL_ROS_TIME) {
+        bool timer_override_active = false;
+        if (rcl_is_enabled_ros_time_override(clock, &timer_override_active) != RCL_RET_OK) {
+          // should never happen
+          return RCL_RET_ERROR;
+        }
+
+        if (timer_override_active) {
+          // we need to check, it the timer is already ready
+          bool override_timer_is_ready = false;
+          if (rcl_timer_is_ready(wait_set->timers[t_idx], &override_timer_is_ready) != RCL_RET_OK) {
+            // should never happen
+            return RCL_RET_ERROR;
+          }
+
+          if (override_timer_is_ready) {
+            // no need to search further for the timeout, we need to wake up instantly
+            is_non_blocking = true;
+            break;
+          }
+
+          // if the timer override is active, there is no point in computing a wait time,
+          // as it might be on a total wrong time basis. In case this timer becomes ready,
+          // the guard_condition above will wake us.
+          continue;
+        }
+      }
+
+      // get the time of the next call to the timer
+      int64_t next_call_time = INT64_MAX;
+      rcl_ret_t ret = rcl_timer_get_next_call_time(wait_set->timers[t_idx], &next_call_time);
       if (ret == RCL_RET_TIMER_CANCELED) {
-        wait_set->timers[i] = NULL;
+        wait_set->timers[t_idx] = NULL;
         continue;
       }
       if (ret != RCL_RET_OK) {
         return ret;  // The rcl error state should already be set.
       }
-      if (timer_timeout < min_timeout) {
-        is_timer_timeout = true;
-        min_timeout = timer_timeout;
+      if (next_call_time < min_next_call_time[clock->type]) {
+        clocks[clock->type] = clock;
+        min_next_call_time[clock->type] = next_call_time;
       }
     }
   }
 
-  if (timeout == 0) {
-    // Then it is non-blocking, so set the temporary storage to 0, 0 and pass it.
+  if (is_non_blocking) {
     temporary_timeout_storage.sec = 0;
     temporary_timeout_storage.nsec = 0;
     timeout_argument = &temporary_timeout_storage;
-  } else if (timeout > 0 || is_timer_timeout) {
+  } else {
+    int64_t min_timeout = timeout > 0 ? timeout : INT64_MAX;
+    bool has_valid_timeout = timeout > 0;
+
+    // determine the min timeout of all clocks
+    for (size_t i = RCL_ROS_TIME; i <= RCL_STEADY_TIME; i++) {
+      if (clocks[i] == 0) {
+        continue;
+      }
+
+      int64_t cur_time;
+      rmw_ret_t ret = rcl_clock_get_now(clocks[i], &cur_time);
+      if (ret != RCL_RET_OK) {
+        return ret;  // The rcl error state should already be set.
+      }
+
+      int64_t timer_timeout = min_next_call_time[i] - cur_time;
+
+      if (timer_timeout <= min_timeout) {
```

Comment:
Is it safe to compare timeouts belonging to different clocks?
If I have a wall-time clock that is ready in 10 seconds and a sim-time clock that is ready in 20 seconds, the real-time factor plays an important role in how much I should sleep.

Similarly, and more difficult to evaluate, this could happen when comparing system clock and steady clock as the first could jump back and forth in time.

Reply:
At this point, you won't get any timeout for an override timer. As soon as override is active, we don't compute a timeout, as we rely on the guard conditions to wake us up. 

>If I have a wall-time clock that is ready in 10 seconds and a sim-time clock that is ready in 20 seconds, the real-time factor plays an important role in how much I should sleep.
(e.g. if rtf<= 2, then I should sleep for 10 seconds (wall time), but if rtf >2 then I should sleep for 20 / rtf seconds)

This exactly the main issue, this PR fixes.

Time jumps of the system time are in general a problem, but this is out of the scope of this PR